shared-lib/spiutils/src/protocol/flash.rs

```diff
@@ -0,0 +1,264 @@
+        ReadJedec = 0x9f,
+
+        /// Retrieves SFDP as configured in sfdp registers.
+        /// Implemented in hardware.
+        ReadSfdp = 0x5a,
+
+        ////////////////////////////////////////////////////////////
+        // Read commands
+
+        /// Retrieves data. The behavior of this command depends on the selected
+        /// mode.
+        NormalRead = 0x03,
+
+        /// Retrieves data. The behavior of this command depends on the selected
+        /// mode. Fast read includes a 1 byte delay after retrieving the last
+        /// bit of the addrees before the first bit of data is delivered.
+        FastRead = 0x0b,
+
+        /// Similar to FastReads but uses explicit 4 byte addressing.
+        FastRead4B = 0x0c,
+
+        /// Similar to FastRead with output on both MISO and MOSI.
+        FastReadDualOutput = 0x3b,
+
+        ////////////////////////////////////////////////////////////
+        // Address mode commands
+
+        /// Enable 4 byte address mode.
+        /// Must be implemented in software.
+        Enter4ByteAddressMode = 0xb7,
+
+        /// Disable 4 byte address mode and revert to 3 byte address mode.
+        /// Must be implemented in software.
+        Exit4ByteAddressMode = 0xe9,
+    }
+}
+
+impl OpCode {
+    /// Returns true iff the OpCode requires an address.
+    pub fn has_address(&self) -> bool {
+        match self {
+            Self::SectorErase => true,
+            Self::BlockErase32KB => true,
+            Self::BlockErase64KB => true,
+            Self::PageProgram => true,
+            Self::ReadSfdp => true,
+            Self::NormalRead => true,
+            Self::FastRead => true,
+            Self::FastRead4B => true,
+            Self::FastReadDualOutput => true,
+            _ => false,
+        }
+    }
+
+    /// Returns true iff the OpCode requires a dummy byte.
+    pub fn has_dummy_byte(&self) -> bool {
+        match self {
+            Self::FastRead => true,
+            Self::FastRead4B => true,
+            Self::FastReadDualOutput => true,
+            _ => false,
+        }
+    }
+
+    /// Returns true iff the OpCode requires data.
+    pub fn has_data(&self) -> bool {
+        match self {
+            Self::PageProgram => true,
+            _ => false,
+        }
+    }
+}
+
+const DUMMY_BYTE_VALUE: u8 = 0xff;
+
+/// A parsed SPI flash protocol header.
+#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug)]
+pub struct Header<AddrType> {
+    /// The SPI op code
+    pub opcode: OpCode,
+
+    /// The address.
+    ///
+    /// Note that not all SPI op codes require an address.
+    pub address: Option<AddrType>,
+}
+
+impl<'a, AddrType> Header<AddrType>
+where AddrType: BeInt {
+    fn from_wire<R: Read<'a>>(mut r: R) -> Result<Self, FromWireError> {
+        let opcode_u8 = r.read_be::<u8>()?;
+        let opcode = OpCode::from_wire_value(opcode_u8).ok_or(FromWireError::OutOfRange)?;
+
+        let mut address : Option<AddrType> = None;
```

Comment:
Formatting nitpick: remove the space before the `:` character.

Reply:
Done.